evm_arithmetization/src/generation/state.rs

```diff
@@ -1,26 +1,216 @@
 use std::collections::HashMap;
 
+use anyhow::bail;
 use ethereum_types::{Address, BigEndianHash, H160, H256, U256};
 use keccak_hash::keccak;
+use log::log_enabled;
 use plonky2::field::types::Field;
 
 use super::mpt::{load_all_mpts, TrieRootPtrs};
 use super::TrieInputs;
 use crate::cpu::kernel::aggregator::KERNEL;
 use crate::cpu::kernel::constants::context_metadata::ContextMetadata;
+use crate::cpu::membus::NUM_GP_CHANNELS;
+use crate::cpu::stack::MAX_USER_STACK_SIZE;
 use crate::generation::rlp::all_rlp_prover_inputs_reversed;
+use crate::generation::CpuColumnsView;
 use crate::generation::GenerationInputs;
 use crate::memory::segments::Segment;
 use crate::util::u256_to_usize;
 use crate::witness::errors::ProgramError;
-use crate::witness::memory::{MemoryAddress, MemoryState};
+use crate::witness::memory::MemoryChannel::GeneralPurpose;
+use crate::witness::memory::{MemoryAddress, MemoryOp, MemoryState};
+use crate::witness::memory::{MemoryOpKind, MemorySegmentState};
+use crate::witness::operation::{generate_exception, Operation};
 use crate::witness::state::RegistersState;
 use crate::witness::traces::{TraceCheckpoint, Traces};
-use crate::witness::util::stack_peek;
+use crate::witness::transition::{
+    decode, fill_op_flag, get_op_special_length, log_kernel_instruction, might_overflow_op,
+    read_code_memory, Transition,
+};
+use crate::witness::util::{
+    fill_channel_with_value, mem_read_gp_with_log_and_fill, stack_peek, stack_pop_with_log_and_fill,
+};
 
-pub(crate) struct GenerationStateCheckpoint {
-    pub(crate) registers: RegistersState,
-    pub(crate) traces: TraceCheckpoint,
+/// A State is either an `Interpreter` (used for tests and jumpdest analysis) or
+/// a `GenerationState`.
+pub(crate) trait State<F: Field> {
+    /// Returns a `State`'s `Checkpoint`.
```

Comment:
```suggestion
    /// Returns a mutable reference to the `State`'s latest `Checkpoint`.
```

Reply:
But it doesn't return a mutable reference to the checkpoint, right? The `mut` is necessary because we reset the memory_ops to the empty vector in the interpreter.